test/integration/http_integration.cc

```diff
@@ -2117,6 +2117,287 @@ void HttpIntegrationTest::testTrailers(uint64_t request_size, uint64_t response_
   }
 }
 
+// Adds metadata inserting filter before metadata consuming filter. Verify metadata are consumed.
+void HttpIntegrationTest::testInsertBeforeConsumeResponseMetadata() {
+  config_helper_.addFilter(R"EOF(
+name: response-metadata-insert-filter
+config: {}
+)EOF");
+
+  config_helper_.addFilter(R"EOF(
+name: response-metadata-consume-filter
+config: {}
+)EOF");
+
+  config_helper_.addConfigModifier(
+      [&](envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager& hcm)
+          -> void { hcm.set_proxy_100_continue(true); });
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Upstream responds with headers.
+  auto response = codec_client_->makeRequestWithBody(default_request_headers_, 10);
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+
+  // Verifies metadata added in inserting filter are passed to the clients, except the ones consumed
+  // by the consuming filter.
+  response->waitForEndStream();
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ(response->metadata_map().find("headers")->second, "headers");
+  EXPECT_EQ(response->metadata_map().find("duplicate")->second, "duplicate");
+  EXPECT_EQ(response->metadata_map().count("metadata"), 0);
+  EXPECT_EQ(response->metadata_map().size(), 2);
+
+  // Upstream responds with headers and data.
+  response = codec_client_->makeRequestWithBody(default_request_headers_, 10);
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+  upstream_request_->encodeData(100, true);
+
+  // Verifies metadata at client.
```

Comment:
I think if it would be helpful for these to replace the existing comment with what's actually being tested

// Verifies metadata at client. ->
// Verifies that encodeData above results in "data" being added to the metadata map.

Also if all the tests are extending previous metadata entries, maybe just declare the map once, add to it for each code block, and have a function verifyExpectedMetadata so we don't have to copy/paste "headers" "duplicate" etc... 7 times?

Reply:
Good idea! Fixed.